### PR TITLE
override defaults for containers

### PR DESCRIPTION
**Pull request by gorkem**
Overrides default JVM values when started on a container.

### CODE REVIEW

File: src/javaServerStarter.ts [hunk 1]
```diff
@@ -3,16 +3,18 @@ import * as path from 'path';
...
 import * as glob from 'glob';
 import * as os from 'os';
+import isDocker = require('is-Docker');
 import { StreamInfo, Executable, ExecutableOptions } from 'vscode-languageclient';
 import { RequirementsData } from './requirements';
 import { getJavaEncoding } from './settings';
 
 declare var v8debug;
 const DEBUG = (typeof v8debug === 'object') || startedInDebugMode();
+const DEFAULT_VM_ARGS = '-noverify -Xmx1G -XX:+UseG1GC -XX:+UseStringDeduplication';
```
Comment (fbricon) [inline src/javaServerStarter.ts]: I'm not sure if there's a way to get the default pref value from the VS Code API, but if you can't, you can instead read the default value from the package.json directly with something like 
     let extensionPackage = require(ExtensionContext.asAbsolutePath('./package.json'));

File: src/javaServerStarter.ts [hunk 2]
```diff
@@ -61,6 +63,9 @@ function prepareParams(requirements: RequirementsData, javaConfiguration, worksp
 	}
 
 	let vmargs = javaConfiguration.get('jdt.ls.vmargs', '');
+	if( DEFAULT_VM_ARGS === vmargs && isDocker() ) {
+		vmargs = '-XX:+UseContainerSupport -XX:MaxRAMPercentage=80';
```
Comment (fbricon) [inline src/javaServerStarter.ts]: you're losing -noverify -XX:+UseG1GC -XX:+UseStringDeduplication

File: src/javaServerStarter.ts [hunk 3]
```diff
@@ -61,6 +63,9 @@ function prepareParams(requirements: RequirementsData, javaConfiguration, worksp
 	}
 
 	let vmargs = javaConfiguration.get('jdt.ls.vmargs', '');
+	if( DEFAULT_VM_ARGS === vmargs && isDocker() ) {
```
Comment (fbricon) [inline src/javaServerStarter.ts]: DEFAULT_VM_ARGS = javaConfiguration.inspect('jdt.ls.vmargs').defaultValue

File: src/javaServerStarter.ts (hunk 1 above)
Comment (fbricon) [reply src/javaServerStarter.ts]: See WorkspaceConfiguration.inspect(key).defaultValue

**Comment by snjeza** [comment]
"-XX:+UseContainerSupport -XX:MaxRAMPercentage=80" requires JDK 10.
Che-7 uses JDK 8 - https://github.com/eclipse/che-plugin-registry/blob/master/plugins/org.eclipse.che.vscode-redhat.java/0.38.0/meta.yaml

**Comment by gorkem** [comment]
@snjeza  -XX:+UseContainerSupport has been back ported to java8u191, [changelog](https://www.oracle.com/technetwork/java/javase/8u191-relnotes-5032181.html)

**Comment by snjeza** [comment]
@gorkem I have added '-XX:+UseContainerSupport -XX:MaxRAMPercentage=80' to JDK 1.8.0_201 and got the following error:
```
Improperly specified VM option 'MaxRAMPercentage=80'
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.
```

**Comment by snjeza** [comment]
JDK 1.8.0_201 recognizes  '-XX:+UseContainerSupport -XX:MaxRAMPercentage=80.0' . See https://bugs.java.com/bugdatabase/view_bug.do?bug_id=8219312

**Comment by gorkem** [comment]
closing

**Pull request closed without merging**